File: db/user_repository.py
import json
from typing import Optional
from db.database import get_pool
from db.schema import DEFAULT_PROFILE
from utils.logger import get_logger

log = get_logger("UserRepository")


# ── Users ──────────────────────────────────────────────────────────────────

async def find_or_create_user(session_id: str) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT * FROM users WHERE session_id = $1", session_id
        )
        if row:
            return dict(row)
        row = await conn.fetchrow(
            "INSERT INTO users (session_id) VALUES ($1) RETURNING *", session_id
        )
        log.info(f"New user created: {session_id}")
        return dict(row)


# ── Semantic profile ────────────────────────────────────────────────────────

async def get_profile(user_id: str) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            "SELECT profile FROM user_profiles WHERE user_id = $1", user_id
        )
        if row:
            return json.loads(row["profile"]) if isinstance(row["profile"], str) else dict(row["profile"])
        return dict(DEFAULT_PROFILE)


async def upsert_profile(user_id: str, updates: dict) -> None:
    pool = await get_pool()
    async with pool.acquire() as conn:
        await conn.execute(
            """
            INSERT INTO user_profiles (user_id, profile)
            VALUES ($1, $2::jsonb)
            ON CONFLICT (user_id)
            DO UPDATE SET
                profile    = user_profiles.profile || $2::jsonb,
                updated_at = NOW()
            """,
            user_id,
            json.dumps(updates),
        )
    log.debug(f"Profile updated for {user_id}: {updates}")


async def append_to_profile_array(user_id: str, field: str, value: str) -> None:
    pool = await get_pool()
    async with pool.acquire() as conn:
        await conn.execute(
            f"""
            UPDATE user_profiles
            SET profile = jsonb_set(
                profile,
                '{{{field}}}',
                (COALESCE(profile->'{field}', '[]'::jsonb) || $1::jsonb)
            ),
            updated_at = NOW()
            WHERE user_id = $2
            """,
            json.dumps(value),
            user_id,
        )
    log.debug(f"Appended to profile.{field} for {user_id}")


# ── Prompt formatting ───────────────────────────────────────────────────────

def format_profile_for_prompt(profile: dict) -> str:
    lines = []
    if profile.get("preferred_city"):
        lines.append(f"- Preferred delivery city: {profile['preferred_city']}")
    if profile.get("currency"):
        lines.append(f"- Preferred currency: {profile['currency']}")
    bmin, bmax = profile.get("budget_min"), profile.get("budget_max")
    if bmin is not None or bmax is not None:
        currency = profile.get("currency", "LKR")
        lines.append(f"- Typical budget: {bmin or 'any'} – {bmax or 'any'} {currency}")
    if profile.get("past_recipients"):
        lines.append(f"- Known recipients: {', '.join(profile['past_recipients'])}")
    if profile.get("dislikes"):
        lines.append(f"- Dislikes / avoid: {', '.join(profile['dislikes'])}")
    if profile.get("interests"):
        lines.append(f"- Interests: {', '.join(profile['interests'])}")

    return "User profile:\n" + "\n".join(lines) if lines else "No profile data available yet."

File: db/userRepository.js
import pg from "pg";
import config from "../config/index.js";
import { createLogger } from "../utils/logger.js";
import { DEFAULT_PROFILE } from "./schema.js";

const log = createLogger("UserRepository");
const pool = new pg.Pool({ connectionString: config.db.url });

// ─────────────────────────────────────────
// User (identity)
// ─────────────────────────────────────────

export async function findOrCreateUser(sessionId) {
  const existing = await pool.query(
    `SELECT * FROM users WHERE session_id = $1`,
    [sessionId]
  );
  if (existing.rows.length > 0) return existing.rows[0];

  const created = await pool.query(
    `INSERT INTO users (session_id) VALUES ($1) RETURNING *`,
    [sessionId]
  );
  log.info("Created new user", { sessionId });
  return created.rows[0];
}

export async function getUserBySessionId(sessionId) {
  const res = await pool.query(
    `SELECT * FROM users WHERE session_id = $1`,
    [sessionId]
  );
  return res.rows[0] ?? null;
}

// ─────────────────────────────────────────
// Semantic profile
// ─────────────────────────────────────────

export async function getProfile(userId) {
  const res = await pool.query(
    `SELECT profile FROM user_profiles WHERE user_id = $1`,
    [userId]
  );
  return res.rows[0]?.profile ?? { ...DEFAULT_PROFILE };
}

export async function upsertProfile(userId, updates) {
  // Merge updates into existing profile (JSONB concat operator ||)
  await pool.query(
    `INSERT INTO user_profiles (user_id, profile)
     VALUES ($1, $2)
     ON CONFLICT (user_id)
     DO UPDATE SET profile = user_profiles.profile || $2::jsonb,
                   updated_at = NOW()`,
    [userId, JSON.stringify(updates)]
  );
  log.debug("Profile updated", { userId, updates });
}

// Append to an array field in the profile (e.g. past_recipients, interests)
export async function appendToProfileArray(userId, field, value) {
  await pool.query(
    `UPDATE user_profiles
     SET profile = jsonb_set(
       profile,
       $1,
       (COALESCE(profile->$2, '[]'::jsonb) || $3::jsonb)
     ),
     updated_at = NOW()
     WHERE user_id = $4`,
    [`{${field}}`, field, JSON.stringify(value), userId]
  );
  log.debug(`Appended to profile.${field}`, { userId, value });
}

// ─────────────────────────────────────────
// Profile summary string (injected into prompts)
// ─────────────────────────────────────────

export function formatProfileForPrompt(profile) {
  const lines = [];

  if (profile.preferred_city)
    lines.push(`- Preferred delivery city: ${profile.preferred_city}`);

  if (profile.currency)
    lines.push(`- Preferred currency: ${profile.currency}`);

  if (profile.budget_min != null || profile.budget_max != null) {
    const min = profile.budget_min ?? "any";
    const max = profile.budget_max ?? "any";
    lines.push(`- Typical budget: ${min} – ${max} ${profile.currency ?? "LKR"}`);
  }

  if (profile.past_recipients?.length)
    lines.push(`- Known recipients: ${profile.past_recipients.join(", ")}`);

  if (profile.dislikes?.length)
    lines.push(`- Dislikes / avoid: ${profile.dislikes.join(", ")}`);

  if (profile.interests?.length)
    lines.push(`- Interests: ${profile.interests.join(", ")}`);

  return lines.length
    ? `User profile:\n${lines.join("\n")}`
    : "No profile data available yet.";
}

export { pool };
